toupee/ensembles/methods.py

```diff
@@ -202,4 +219,78 @@ def _on_model_end(self):
 
     def _on_model_start(self):
         """ Default callback when a model starts training """
-        pass
\ No newline at end of file
+        self.data = self.data.resample()
+
+
+class DynamicMembers(EnsembleMethod):
+    """
+    A simple Ensemble - repeat the training N times and aggregate the results
+    """
+    def _initialise_members(self):
+        pass
+
+    def _members(self):
+        """ Must be implemented by the inheriting class """
+        raise NotImplementedError()
+
+
+class DIB(DynamicMembers):
+    """
+    Deep Incremental Boosting ()
+    """
+    def __init__(self, subsequent_epochs, insert_after, new_layers, variant='M1', **kwargs):
+        super().__init__(aggregator='averaging', **kwargs)
+        self.subsequent_epochs = subsequent_epochs
+        self.new_layers = new_layers
+        self.insert_after = insert_after
+        self.variant = variant
+        self.sample_weights = np.ones(self.data.size['train']) / float(self.data.size['train'])
+        self.model_weights = np.ones(self.size)
+
+    def _on_model_end(self):
+        """ Default callback when a model finishes training """
+        model = self._fit_loop_info['current_model']
+        y_true = []
+        y_pred_p = []
+        for (x, y_true_batch) in self.data.get_training_handle(resample=False): #TODO: non-resampled handle here
+            y_true.append(np.argmax(y_true_batch, axis=1))
+            y_pred_p.append(model.predict_proba(x))
+        y_true = np.concatenate(y_true)
+        y_pred_p = np.concatenate(y_pred_p)
+        y_pred = np.argmax(y_pred_p, axis=1)
+        if self.variant == 'MA': 
+            y_pred_weights = np.max(y_pred_p, axis=1)
+        else:
+            y_pred_weights = 1.
+        errors = (y_true == y_pred).astype('int32')
+        e = np.sum(errors * self.sample_weights * y_pred_weights)
```

Comment:
`e` can be confused with Euler's Number, can we get a better variable name? :D

Reply:
Well, it's called epsilon in the literature, and that is clearly way too many keystrokes 😆